### ISSUE TITLE
BUG: parameters not being updated properly

### ISSUE THREAD

**Issue by andyfaff** [lmfit/lmfit-py#265]
With current master there is a bug when setting parameter values, if there are constraints and limits.  Take the following example below.
`a` is always supposed to be twice `b` and it has no limits. `b` is clamped between 0 and 20.  Whenever I set `b` to a new value, and `b` is within its limits, `a` is recalculated properly. If I set `b` outside it's limits, then `b` is clamped to `b.max` correctly. It follows that `a` should be twice `b.max`.  However, this isn't the case.  I have a fix for this, it'll be included in #261

    import numpy as np
    from lmfit import Parameter, Parameters
    from numpy.testing import assert_, assert_almost_equal

    p = Parameters()

    p['a'] = Parameter('a', 10, True)
    p['b'] = Parameter('b', 10, True, 0, 20)

    p['a'].expr = '2 * b'
    assert_almost_equal(p['a'].value, 20)

    p['b'].value = 15
    assert_almost_equal(p['b'].value, 15)
    assert_almost_equal(p['a'].value, 30)

    p['b'].value = 30
    assert_almost_equal(p['b'].value, 20)
    assert_almost_equal(p['a'].value, 40)

    ---------------------------------------------------------------------------
    AssertionError                            Traceback (most recent call last)
    <ipython-input-6-dd8c6e7c13e7> in <module>()
          1 p['b'].value = 30
          2 assert_almost_equal(p['b'].value, 20)
    ----> 3 assert_almost_equal(p['a'].value, 40)

    /Users/anz/miniconda3/envs/dev3/lib/python3.4/site-packages/numpy/testing/utils.py in assert_almost_equal(actual, desired, decimal, err_msg, verbose)
        508         pass
        509     if round(abs(desired - actual), decimal) != 0:
    --> 510         raise AssertionError(_build_err_msg())
        511 
        512 

    AssertionError: 
    Arrays are not almost equal to 7 decimals
     ACTUAL: 60
     DESIRED: 40

**Comment by andyfaff** [comment]
I'll also include this test in the PR

**Comment by newville** [comment]
@andyfaff   Yes,  you have to run `p.update_constraints()` to guarantee that constrained parameters have up-to-date values.  That is potentially expensive, so within a fitting loop, it doesn't really make sense to run that on every assignment -- it just needs to be run once after assigning all variable parameters values.  Since "within a fitting loop" is the main point, that is the default behavior.

Although, now that I look, I see that `update_constraints()` was removed from `Minimizer.__residual()` in 352be5fc9 (Mar 1st) when asteval was moved from `Minimizer` to `Parameters`.    That is definitely a mistake (though I suspect won't alter final results as the values stabilize in the final iterations), and I will fix this today.

**Comment by newville** [comment]
@andyfaff Over at the discussion of #266 you said including a call to `update_constraints` still fails.  I suspect something else is wrong.  This works for me:

```python
from lmfit.parameter import Parameter, Parameters
from numpy.testing import assert_, assert_almost_equal

pars = Parameters()

pars.add('b', value=10, vary=True, min=0, max=20.0)
pars.add('a', expr='2*b')


pars['b'].value = 15.0
assert_almost_equal(pars['b'].value, 15)
assert_almost_equal(pars['a'].value, 2*pars['b'].value)

print(pars.valuesdict())
print('====================')

pars['b'].value = 99.0
pars.update_constraints()

assert_almost_equal(pars['b'].value, 20.0)
assert_almost_equal(pars['a'].value, 2*pars['b'].value)
print(pars.valuesdict())
```

**Comment by newville** [comment]
addressed in 208871db96689c911e00b9d59dace2de68df3220
